File: platform/polyglot/saucelabs_api/mjs/src/rate-limiter.mjs
/**
 * Rate Limiter Module — Sauce Labs API Client
 *
 * Reactive rate limiter for the Sauce Labs REST API.
 * Sauce Labs allows ~10 req/sec (3,500/hour) for authenticated requests.
 * HTTP 429 + Retry-After header signals rate limits.
 *
 * This limiter reacts to 429 responses with exponential backoff:
 *   - Parses the Retry-After header
 *   - Applies exponential backoff with jitter
 *   - Optionally auto-waits and retries
 *   - Invokes an onRateLimit callback for visibility
 */

import { create } from './logger.mjs';
import { SaucelabsRateLimitError } from './errors.mjs';

const log = create('saucelabs-api', import.meta.url);

/**
 * Parse the Retry-After header value into seconds.
 * Handles both numeric seconds and HTTP-date formats.
 *
 * @param {string|null|undefined} headerValue
 * @returns {number} Seconds to wait (defaults to 60 if unparseable)
 */
export function parseRetryAfter(headerValue) {
  if (!headerValue) return 60;

  const numeric = parseFloat(headerValue);
  if (!Number.isNaN(numeric) && numeric > 0) {
    return Math.ceil(numeric);
  }

  const dateMs = Date.parse(headerValue);
  if (!Number.isNaN(dateMs)) {
    const diffSeconds = Math.ceil((dateMs - Date.now()) / 1000);
    return diffSeconds > 0 ? diffSeconds : 1;
  }

  return 60;
}

/**
 * Build a RateLimitInfo object from response headers.
 *
 * @param {object} headers - Response headers (plain object, lowercase keys)
 * @returns {import('./types.mjs').RateLimitInfo}
 */
export function buildRateLimitInfo(headers) {
  const retryAfterRaw = headers['retry-after'] ?? headers['Retry-After'];
  const retryAfter = parseRetryAfter(retryAfterRaw);

  const remaining = headers['x-ratelimit-remaining'] != null
    ? parseInt(headers['x-ratelimit-remaining'], 10)
    : null;

  const limit = headers['x-ratelimit-limit'] != null
    ? parseInt(headers['x-ratelimit-limit'], 10)
    : null;

  const resetRaw = headers['x-ratelimit-reset'];
  const resetAt = resetRaw ? new Date(parseInt(resetRaw, 10) * 1000) : null;

  return { retryAfter, remaining, limit, resetAt, timestamp: new Date() };
}

/**
 * Calculate exponential backoff with jitter.
 *
 * @param {number} retryCount - Current retry attempt (0-indexed)
 * @param {number} [baseDelay=1] - Base delay in seconds
 * @param {number} [maxDelay=60] - Maximum delay cap in seconds
 * @returns {number} Delay in seconds
 */
export function calculateBackoff(retryCount, baseDelay = 1, maxDelay = 60) {
  const exponential = baseDelay * Math.pow(2, retryCount);
  const jitter = Math.random() * baseDelay;
  return Math.min(exponential + jitter, maxDelay);
}

function _sleep(seconds) {
  return new Promise((resolve) => setTimeout(resolve, seconds * 1000));
}

/**
 * Reactive rate limiter for the Sauce Labs REST API.
 */
export class RateLimiter {
  constructor({ autoWait = true, maxRetries = 5, onRateLimit = null, logger = null } = {}) {
    this._autoWait = autoWait;
    this._maxRetries = maxRetries;
    this._onRateLimit = onRateLimit;
    this._logger = logger || log;
    this.lastRateLimit = null;
  }

  /**
   * Handle an HTTP 429 response.
   *
   * @param {Response} response - The fetch Response object
   * @param {function(): Promise<*>} retryFn - Retry callback
   * @param {number} [retryCount=0] - Current retry attempt
   * @returns {Promise<*>}
   * @throws {SaucelabsRateLimitError}
   */
  async handleResponse(response, retryFn, retryCount = 0) {
    const headers = {};
    if (response.headers && typeof response.headers.forEach === 'function') {
      response.headers.forEach((value, key) => { headers[key.toLowerCase()] = value; });
    } else if (response.headers && typeof response.headers === 'object') {
      for (const [k, v] of Object.entries(response.headers)) { headers[k.toLowerCase()] = v; }
    }

    const rateLimitInfo = buildRateLimitInfo(headers);
    this.lastRateLimit = rateLimitInfo;

    this._logger.warn('rate limited by Sauce Labs API (HTTP 429)', {
      retryAfter: rateLimitInfo.retryAfter,
      retryCount,
      maxRetries: this._maxRetries,
      remaining: rateLimitInfo.remaining,
    });

    // Invoke callback
    if (this._onRateLimit) {
      const callbackResult = this._onRateLimit(rateLimitInfo);
      if (callbackResult === false) {
        this._logger.info('onRateLimit callback returned false, aborting retry');
        let body;
        try { body = await response.json(); } catch { body = null; }
        throw new SaucelabsRateLimitError('Rate limit exceeded (callback declined retry)', {
          retryAfter: rateLimitInfo.retryAfter, responseBody: body, headers,
        });
      }
    }

    // Check if we should retry
    if (!this._autoWait || retryCount >= this._maxRetries) {
      const reason = !this._autoWait
        ? 'auto-wait disabled'
        : `max retries exhausted (${retryCount}/${this._maxRetries})`;
      this._logger.warn(`not retrying: ${reason}`);
      let body;
      try { body = await response.json(); } catch { body = null; }
      throw new SaucelabsRateLimitError(`Rate limit exceeded (${reason})`, {
        retryAfter: rateLimitInfo.retryAfter, responseBody: body, headers,
      });
    }

    // Use Retry-After if present, otherwise exponential backoff
    const waitSeconds = rateLimitInfo.retryAfter > 0
      ? rateLimitInfo.retryAfter
      : calculateBackoff(retryCount);

    this._logger.info(`waiting ${waitSeconds}s before retry ${retryCount + 1}/${this._maxRetries}`);
    await _sleep(waitSeconds);
    return retryFn();
  }
}

File: platform/polyglot/saucelabs_api/py/saucelabs_api/rate_limiter.py
"""
Reactive rate limiter for the Sauce Labs API client.

Handles HTTP 429 responses by parsing ``Retry-After``, optionally sleeping
with exponential backoff, and retrying the original request.
"""

from __future__ import annotations

import asyncio
import inspect
import math
import random
from datetime import datetime, timezone
from typing import Any, Awaitable, Callable

from .errors import SaucelabsRateLimitError, create_error_from_response
from .logger import _Logger, create_logger
from .types import RateLimitInfo


class RateLimiter:
    """Reactive rate limiter for the Sauce Labs REST API.

    Parameters
    ----------
    auto_wait:
        If ``True``, the limiter sleeps and retries on HTTP 429.
    max_retries:
        Maximum consecutive retries before raising.
    on_rate_limit:
        Optional callback invoked with :class:`RateLimitInfo` on 429.
    logger:
        Custom logger instance.
    """

    __slots__ = (
        "_auto_wait",
        "_max_retries",
        "_on_rate_limit",
        "_logger",
        "_last_rate_limit",
    )

    def __init__(
        self,
        *,
        auto_wait: bool = True,
        max_retries: int = 5,
        on_rate_limit: (
            Callable[[RateLimitInfo], Awaitable[bool]]
            | Callable[[RateLimitInfo], bool]
            | None
        ) = None,
        logger: _Logger | Any | None = None,
    ) -> None:
        self._auto_wait = auto_wait
        self._max_retries = max_retries
        self._on_rate_limit = on_rate_limit
        self._logger: Any = logger or create_logger("saucelabs_api", "rate_limiter")
        self._last_rate_limit: RateLimitInfo | None = None

    @property
    def last_rate_limit(self) -> RateLimitInfo | None:
        return self._last_rate_limit

    @staticmethod
    def _parse_retry_after(headers: dict[str, str]) -> float:
        raw = headers.get("retry-after") or headers.get("Retry-After", "")
        try:
            return max(float(raw), 0.1)
        except (ValueError, TypeError):
            return 1.0

    @staticmethod
    def _calculate_backoff(retry_count: int, base_delay: float = 1.0, max_delay: float = 60.0) -> float:
        exponential = base_delay * math.pow(2, retry_count)
        jitter = random.random() * base_delay
        return min(exponential + jitter, max_delay)

    def _build_info(self, headers: dict[str, str]) -> RateLimitInfo:
        retry_after = self._parse_retry_after(headers)
        remaining_raw = headers.get("x-ratelimit-remaining")
        limit_raw = headers.get("x-ratelimit-limit")
        return RateLimitInfo(
            retry_after=retry_after,
            remaining=int(remaining_raw) if remaining_raw else None,
            limit=int(limit_raw) if limit_raw else None,
            reset_at=headers.get("x-ratelimit-reset"),
            timestamp=datetime.now(timezone.utc).isoformat(),
        )

    async def handle_response(
        self,
        response: Any,
        retry_fn: Callable[[], Awaitable[Any]],
        retry_count: int = 0,
    ) -> Any:
        """Process an HTTP response and handle 429 status codes."""
        if response.status_code != 429:
            return response

        headers = dict(response.headers)
        info = self._build_info(headers)
        self._last_rate_limit = info

        self._logger.warning(
            f"Rate limited (429). Retry-After: {info.retry_after}s",
            {"retry_count": retry_count, "retry_after": info.retry_after},
        )

        if self._on_rate_limit is not None:
            result = self._on_rate_limit(info)
            if inspect.isawaitable(result):
                result = await result
            if result is False:
                body = response.text if hasattr(response, "text") else ""
                raise create_error_from_response(429, body, headers)

        if not self._auto_wait:
            body = response.text if hasattr(response, "text") else ""
            raise create_error_from_response(429, body, headers)

        if retry_count >= self._max_retries:
            body = response.text if hasattr(response, "text") else ""
            raise create_error_from_response(429, body, headers)

        wait_seconds = info.retry_after if info.retry_after > 0 else self._calculate_backoff(retry_count)

        self._logger.info(
            f"Waiting {wait_seconds}s before retry {retry_count + 1}/{self._max_retries}",
        )
        await asyncio.sleep(wait_seconds)

        new_response = await retry_fn()
        return await self.handle_response(new_response, retry_fn, retry_count + 1)
